trieste/utils/pareto.py

```diff
@@ -101,10 +110,168 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
         )
 
         range_ = tf.range(len_front + 1)[:, None]
-        lower = tf.concat([range_, tf.zeros_like(range_)], axis=-1)
-        upper = tf.concat([range_ + 1, pf_ext_idx[::-1, 1:][: pf_ext_idx[-1, 0]]], axis=-1)
+        lower_result = tf.concat([range_, tf.zeros_like(range_)], axis=-1)
+        upper_result = tf.concat(
+            [range_ + 1, pseudo_front_idx[::-1, 1:][: pseudo_front_idx[-1, 0]]], axis=-1
+        )
+
+        return BoundedVolumes(lower_result, upper_result)
+
+    @staticmethod
+    def _is_test_required(smaller: TensorType) -> TensorType:
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
+
+        return is_dom_augm
+
+    def _divide_conquer_nd(
+        self, front: TensorType, jitter: float, threshold: TensorType | float = 0
+    ) -> TensorType:
+        # Divide and conquer strategy to compute the cells covering the non-dominated region.
+        # Generic version: works for an arbitrary number of objectives.
```

Comment:
Does this has a reference?

Reply:
I'm finding it hard to work out what is going on in this method.

Reply:
the reference is: https://ieeexplore.ieee.org/document/6256586 you can check the description of the algorithm at section: C. C. Branch-and-bound procedure

Reply:
My main source to port/implement this from the gpflowopt code https://github.com/GPflow/GPflowOpt/blob/3d86bcc000b0367f19e9f03f4458f5641e5dde60/gpflowopt/pareto.py#L173

Reply:
Its definitely worth adding the reference in the docs and in the code here. Also maybe refer it to a branch and bound rather than divide an conquer so that users can quickly work out where to read.

Reply:
Its a large block of code and currently only has a single test ( test_pareto_divide_conquer_nd_three_dimension_case). How difficult would it be for you to provide another example (maybe 4 or 5 dimensions)?

Reply:
currently I inspect the validity of the result by visualizing the hyperbounds. for d>3 seems it is not possible to do that (correct me if i am wrong). do you have any suggestion?

Reply:
> 
> 
> Its definitely worth adding the reference in the docs and in the code here. Also maybe refer it to a branch and bound rather than divide an conquer so that users can quickly work out where to read.

Should we put it in the `Pareto` class's docstring? since `_divide_and_conquer` here is private, I am not sure if it is ok to put it in `_divide_and_conquer`

Reply:
Yes in the Pareto class.

Reply:
Once you have this and have brought the branch up to date then I think we are ready to merge! I'll do a final check then :)

Reply:
okay. please check the latest commit. not sure if I did it right.

Reply:
Nice! I'm just going to ask @uri-granta  to give this a quick look over please. It would be good to get an engineers perspective on this! 

We can then merge this PR and https://github.com/secondmind-labs/trieste/pull/217 and focus on the EHVI acquisition function itself.

Reply:
Nice.